onnxruntime/core/providers/rocm/rocm_execution_provider.cc

```diff
@@ -279,14 +279,22 @@ Status ROCMExecutionProvider::OnRunStart() {
   while (it != deferred_release_cpu_ptr_.end()) {
     auto& e = it->first;
     auto& v = it->second;
-    // note that hipEventQuery returns hipSucess before first hipEventRecord
-    if (v.recorded && hipSuccess == hipEventQuery(e)) {
-      for (auto p : v.cpu_ptrs) {
-        cpu_alloc->Free(p);
+    // note that hipEventQuery returns hipSuccess before first hipEventRecord
+    if (v.recorded) {
+      auto event_query_status = hipEventQuery(e);
+      if (event_query_status == hipSuccess) {
+        for (auto p : v.cpu_ptrs) {
+          cpu_alloc->Free(p);
+        }
+        it = deferred_release_cpu_ptr_.erase(it);
+        HIP_RETURN_IF_ERROR(hipEventDestroy(e));
```

Comment:
The CI failed on a segmentation fault. It looks like the reference to _e_ is no longer valid after the erase. Try revert to _hipEvent_t expired_event = it->first;_ to preserve copy of hipEvent_t struct?

Reply:
I was just going through the logs now and saw that, too.  I can either revert the removal of that `expired_event` line, or change `auto& e = it->first` to not be a reference but instead a copy.

Reply:
I guess auto e = it->first is cleaner.

Reply:
One more idea - what if we just swapped the order of these statements?  That way `e` is still valid.
```
        it = deferred_release_cpu_ptr_.erase(it);
        HIP_RETURN_IF_ERROR(hipEventDestroy(e));
```

Reply:
Oh sure, yeah, that should work ✓

Reply:
Fixed.